File: starter/swift/Trans/Temp.swift
//
//  Temp.swift
//  L1 Compiler
//  Temporaries
//  Author: Kaustuv Chaudhuri <kaustuv+@cs.cmu.edu>
//  Modified: Alex Vaynberg <alv@andrew.cmu.edu>
//  Modified: Frank Pfenning <fp@cs.cmu.edu>
//  Converted to OCaml by Michael Duggan <md5i@cs.cmu.edu>
//  Converted to Swift by Jonathan Burns <jaburns@andrew.cmu.edu>
//

struct Temp : CustomStringConvertible, Comparable, Hashable {
    private(set) static var numTemps: Int = 1
    
    let id: Int
    
    init() {
        id = Temp.numTemps
        Temp.numTemps += 1
    }
    
    static func reset() {
        numTemps = 1
    }
    
    var description: String {
        return "%t\(id)"
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

func <(lhs: Temp, rhs: Temp) -> Bool {
    return lhs.id < rhs.id
}

func ==(lhs: Temp, rhs: Temp) -> Bool {
    return lhs.id == rhs.id
}

File: Compiler/lab2/lib/utils/temp.ml
(* L1 Compiler
 * Temporaries
 * Author: Kaustuv Chaudhuri <kaustuv+@cs.cmu.edu>
 * Modified: Alex Vaynberg <alv@andrew.cmu.edu>
 * Modified: Frank Pfenning <fp@cs.cmu.edu>
 * Converted to OCaml by Michael Duggan <md5i@cs.cmu.edu>
 * Modified: Alice Rao <alrao@andrew.cmu.edu>
 *)

open Core

module T =
struct
  type t = I of int | S of string 
  [@@deriving sexp, compare, hash]
end

type t = T.t [@@deriving sexp, compare, hash]

let counter = ref 1
let reset () = counter := 1

let create () =
  let t = T.I !counter in
  incr counter;
  t
;;

let create_with_name (name : string) = 
  T.S name
;;

let name t = match t with | T.I t -> "%t" ^ string_of_int t | S s -> s

include Comparable.Make (T)
